diff --git a/node_modules/@types/flexsearch/index.d.ts b/node_modules/@types/flexsearch/index.d.ts
index ecde8e7..64a5f1e 100755
--- a/node_modules/@types/flexsearch/index.d.ts
+++ b/node_modules/@types/flexsearch/index.d.ts
@@ -6,7 +6,6 @@
 /************************************/
 /* Utils                            */
 /************************************/
-export type Id = number | string;
 export type Limit = number;
 export type ExportHandler<T> = (id: string | number, value: T) => void;
 export type AsyncCallback<T = undefined> = T extends undefined ? () => void : (result: T) => void;
@@ -165,7 +164,7 @@ export type IndexSearchResult = Id[];
  * * Usage: https://github.com/nextapps-de/flexsearch#usage
  */
 
-export class Index {
+export default class Index<ID extends number | string  = number> {
    constructor(x?: Preset | IndexOptions<string>);
    add(id: Id, item: string): this;
    append(id: Id, item: string): this;
